File: 543-diameter-of-binary-tree/diameter-of-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int ans = 0;
    int check(TreeNode* root)
    {
       if(!root) return 0;
       int lh = check(root->left);
       int rh = check(root->right);
       ans = max(ans,lh+rh);
       return 1+max(lh,rh);
    }
    int diameterOfBinaryTree(TreeNode* root) {
       check(root);
       return ans; 
    }
};

File: 543-diameter-of-binary-tree/diameter-of-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int maxi = 0;
    public int check(TreeNode root)
    {
        if(root == null) return 0;
        int lh = check(root.left);
        int rh = check(root.right);
        maxi = Math.max(lh+rh,maxi);
        return 1+Math.max(lh,rh);
    } 
    public int diameterOfBinaryTree(TreeNode root)
    {
        check(root);
        return maxi;
    }
}